projects/memapCore/src/main/java/memap/helper/milp/MILPProblemNoConnections.java

```diff
@@ -273,80 +273,47 @@ public LpSolve createSOCBoundaries(LpSolve problem, BuildingMessage buildingMess
-			if (maxDischargeCapacity >= sm.capacity) {
-				maxDischargeCapacity = sm.capacity;
-			}
-			if (maxDischargeCapacity <= 0) {
-				maxDischargeCapacity = 0.0001;
-			}
-
-			double maxChargeCapacity = sm.capacity - sm.storageEnergyContent;
-			if (maxChargeCapacity <= 0.0) {
-				maxChargeCapacity = 0.0001;
-			}
-			if (maxChargeCapacity >= sm.capacity) {
-				maxChargeCapacity = sm.capacity;
-			}
+			double SOC_perc = sm.stateOfCharge; 
+			double standbyLosses = sm.storageLosses;
 			
-			/*
-			// Hard coded standby losses
-			double standbyLosses = 0.2;
-			// Redeclare the SOC to reflect an actual SOC in decimal percent [0; 1]
-			double SOC_perc = sm.stateOfCharge / sm.capacity; 		
+			// check and enforce that SOC is between 0 and 1, due to numerical issues.
+			if (SOC_perc >= 1) {
+				SOC_perc = 1;
+			}
+			if (SOC_perc <= 0) {
+				SOC_perc = 0;
+			}	
 			
 			// New for SOC within 0 and 1 and standby loss consideration:
 			// helper parameters, only depend on time step length and storage parameters
-			double alpha = 1 - standbyLosses * factor; // Units [-]
-			double beta_to = factor/sm.capacity * sm.efficiencyCharge; // Units [h/kWh]
-			double beta_fm = factor/(sm.capacity * sm.efficiencyDischarge); // Units [h/kWh]
+			double alpha = 1 - standbyLosses * delta_time_factor; // Units [-] 
+			double beta_to = delta_time_factor/sm.capacity * sm.efficiencyCharge; // Units [h/kWh]
+			double beta_fm = delta_time_factor/(sm.capacity * sm.efficiencyDischarge); // Units [h/kWh]			
 			
 			// new temporary constraint rows:
-            double[] rowCHARGE_new = new double[nCols+1];
-            double[] rowDISCHARGE_new = new double[nCols+1];
+            double[] rowCHARGE = new double[nCols+1];
+            double[] rowDISCHARGE = new double[nCols+1];
             
 
             // create new SOC constraints. They are not based on energy/capacity but solely on SOC, i.e. between 0 and 1
             for (int i = 0; i < nStepsMPC; i++) {
             	int index = i + 1 + nStepsMPC * ((controllableHandled * 2)  + volatileHandled + (couplerHandled*2)+ (storageHandled*2)  );  		        	
 	        	// First add the factor for the discharge decision variable (x_fm)
-            	rowCHARGE_new[index] = - beta_fm * Math.pow(alpha, i);
-            	rowDISCHARGE_new[index] = beta_fm * Math.pow(alpha,  i);
+            	rowCHARGE[index] = - beta_fm * Math.pow(alpha, nStepsMPC -1 - i);
+            	rowDISCHARGE[index] = beta_fm * Math.pow(alpha, nStepsMPC -1 - i);
             	
             	// Now add the factor for the charging decision variable x_to:
-            	rowCHARGE_new[index + nStepsMPC] = beta_to * Math.pow(alpha, i);
-            	rowDISCHARGE_new[index + nStepsMPC] = - beta_to * Math.pow(alpha, i);
+            	rowCHARGE[index + nStepsMPC] = beta_to * Math.pow(alpha, nStepsMPC -1 - i);
+            	rowDISCHARGE[index + nStepsMPC] = - beta_to * Math.pow(alpha, nStepsMPC -1 - i);
```

Comment:
That looks very nice, indeed. My only comment concerns the border case. If we assume that "i" starts at 0 and "k" starts at 0, then the loops should be:

`for (int i = 0; i < nStepsMPC; i++) {`

      int index = i + 1 + nStepsMPC * ((controllableHandled * 2)  + volatileHandled + (couplerHandled*2)+ (storageHandled*2)  );`
            	
      double[] rowCHARGE = new double[nCols+1];
      double[] rowDISCHARGE = new double[nCols+1];
    			
      for (int k = 0; k < i; k++) { 
        rowCHARGE[index + k] = - beta_fm * Math.pow(alpha, (i-k) );
        rowDISCHARGE[index + k] = beta_fm * Math.pow(alpha, (i-k) );
    	        		
        rowCHARGE[index + k + nStepsMPC] = beta_to * Math.pow(alpha, (i-1)-k);
        rowDISCHARGE[index + k + nStepsMPC] = - beta_to * Math.pow(alpha, (i-1)-k);
      }
                	
      problem.addConstraint(rowDISCHARGE, LpSolve.LE, (SOC_perc * Math.pow(alpha, i)));
      problem.addConstraint(rowCHARGE, LpSolve.LE, (1-(SOC_perc * Math.pow(alpha, i))));
    }`

also "i" ends at (nStepsMPC - 1), so it should work fine then. Right?

Reply:
That's true. However, then we have to adjust the indexing altogether. The alpha multiplied with the right hand side, i.e. the SOC, has to have an exponent one higher than the largest exponent on the left hand side. Otherwise the charging and discharging gets imposed with losses before it has been stored. 
I propose we simply increase the index for the constraint creation, so that the relative differences in the indices match again and all charging and discharging is considered at one point. 
Also: The `int index` at the start of the i-loop is different now as well. It gets increased by the k and therefore should be:
`int index = indexBuilding + nStepsMPC * ((controllableHandled * 2) + volatileHandled + (couplerHandled * 2) + (storageHandled * 2));`

I tested this snippet and it showed the expected SOC behaviour. Note that the k index now loops until k=1 to make the index shift consistent:
`
 
		        for (int i=0; i<nStepsMPC; i++) {
				int index = indexBuilding + nStepsMPC * ((controllableHandled * 2)  + volatileHandled + (couplerHandled*2)+ (storageHandled*2)  );
				double[] rowCHARGE = new double[nCols+1];
		            	double[] rowDISCHARGE = new double[nCols+1];
		        	for (int k = 0; k <= i; k++) {
		        		rowCHARGE[index + k] = - beta_fm * Math.pow(alpha, i-k);
		        		rowDISCHARGE[index + k] = beta_fm * Math.pow(alpha, i-k);
		        		
		        		rowCHARGE[index + k + nStepsMPC] = beta_to * Math.pow(alpha, i-k);
		            		rowDISCHARGE[index + k + nStepsMPC] = - beta_to * Math.pow(alpha, i-k);
		                }
	            	
		        	problem.addConstraint(rowDISCHARGE, LpSolve.LE, (SOC_perc * Math.pow(alpha, i + 1)));
		        	problem.addConstraint(rowCHARGE, LpSolve.LE, (1-(SOC_perc * Math.pow(alpha, i + 1))));
		        }
`

Reply:
I agree with Johannes on "The alpha multiplied with the right hand side, i.e. the SOC, has to have an exponent one higher than the largest exponent on the left hand side". That's because every charging/discharging setpoint x influences always the SOC of the next step in future and not the current or previous one.

xCharge(0) influences SOC(1)
xCharge(1) influences SOC(2)
...
xCharge(nStepsMPC-1) influences SOC(nStepsMPC)
[xCharge(nStepsMPC) influences SOC(nStepsMPC+1)] - irrelevant

That'a a difference to the former indexing as it was used in the planning tool.

That means, when i starts at 0 and nStepsMPC (e.g. 5 ) variables are to be optimized, then
- xCharge(0)....xCharge(nStepsMPC-1) are optimization variables
- SOC(1)...SOC(nStepsMPC) are to be constrained

Reply:
Cool. Thank for the review. It does all make sense. 

Only thing - I would still keep is the start index. I would keep it as:
`int index = indexBuilding + 1 + nStepsMPC * ((controllableHandled * 2) + volatileHandled + (couplerHandled * 2) + (storageHandled * 2));`

The reason is that our library LPSolve is a little bit akward. It starts counting its arrays at 1 not at 0. This is why our 
`double[] rowCHARGE = new double[nCols+1];`  has one column more than we actually calculate ;-)

It took me a while to find that behaviour. Its is weird. But it can be checked in the *.lp file that we save as well.